File: src/main/java/play/pluv/playlist/controller/PlayListController.java
package play.pluv.playlist.controller;

import static play.pluv.playlist.domain.MusicStreaming.APPLE;
import static play.pluv.playlist.domain.MusicStreaming.SPOTIFY;
import static play.pluv.playlist.domain.MusicStreaming.YOUTUBE;

import jakarta.validation.Valid;
import java.util.List;
import lombok.RequiredArgsConstructor;
import org.springframework.web.bind.annotation.PathVariable;
import org.springframework.web.bind.annotation.PostMapping;
import org.springframework.web.bind.annotation.RequestBody;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RestController;
import play.pluv.base.BaseResponse;
import play.pluv.playlist.application.PlayListService;
import play.pluv.playlist.application.dto.ApplePlayListReadRequest;
import play.pluv.playlist.application.dto.PlayListMusicResponse;
import play.pluv.playlist.application.dto.PlayListOcrRequest;
import play.pluv.playlist.application.dto.PlayListOverViewResponse;
import play.pluv.playlist.application.dto.PlayListReadRequest;

@RestController
@RequestMapping("/playlist")
@RequiredArgsConstructor
public class PlayListController {

  private final PlayListService playListService;

  @PostMapping("/spotify/read")
  public BaseResponse<List<PlayListOverViewResponse>> readSpotifyPlayLists(
      @Valid @RequestBody final PlayListReadRequest request
  ) {
    final var playLists = playListService.getPlayLists(request.getAccessToken(), SPOTIFY);
    final List<PlayListOverViewResponse> response = PlayListOverViewResponse.Companion.createList(playLists);
    return BaseResponse.ok(response);
  }

  @PostMapping("/youtube/read")
  public BaseResponse<List<PlayListOverViewResponse>> readYoutubePlayLists(
      @Valid @RequestBody final PlayListReadRequest request
  ) {
    final var playLists = playListService.getPlayLists(request.getAccessToken(), YOUTUBE);
    final List<PlayListOverViewResponse> response = PlayListOverViewResponse.Companion.createList(playLists);
    return BaseResponse.ok(response);
  }

  @PostMapping("/ocr/read")
  public BaseResponse<List<PlayListMusicResponse>> readOcrPlayLists(
      @Valid @RequestBody final PlayListOcrRequest request
  ) {
    final var musics = playListService.getOcrPlayListMusics(request.getBase64EncodedImages());
    final List<PlayListMusicResponse> responses = PlayListMusicResponse.Companion.createList(musics);
    return BaseResponse.ok(responses);
  }

  @PostMapping("/spotify/{id}/read")
  public BaseResponse<List<PlayListMusicResponse>> readSpotifyMusics(
      @Valid @RequestBody final PlayListReadRequest request, @PathVariable final String id
  ) {
    final var musics = playListService.getPlayListMusics(id, request.getAccessToken(), SPOTIFY);
    final List<PlayListMusicResponse> response = PlayListMusicResponse.Companion.createList(musics);
    return BaseResponse.ok(response);
  }

  @PostMapping("/youtube/{id}/read")
  public BaseResponse<List<PlayListMusicResponse>> readYoutubeMusics(
      @Valid @RequestBody final PlayListReadRequest request, @PathVariable final String id
  ) {
    final var musics = playListService.getPlayListMusics(id, request.getAccessToken(), YOUTUBE);
    final List<PlayListMusicResponse> response = PlayListMusicResponse.Companion.createList(musics);
    return BaseResponse.ok(response);
  }

  @PostMapping("/apple/read")
  public BaseResponse<List<PlayListOverViewResponse>> readApplePlayLists(
      @Valid @RequestBody final ApplePlayListReadRequest request
  ) {
    final var playLists = playListService.getPlayLists(request.getMusicUserToken(), APPLE);
    final List<PlayListOverViewResponse> response = PlayListOverViewResponse.Companion.createList(playLists);
    return BaseResponse.ok(response);
  }

  @PostMapping("/apple/{id}/read")
  public BaseResponse<List<PlayListMusicResponse>> readAppleMusics(
      @Valid @RequestBody final ApplePlayListReadRequest request, @PathVariable final String id
  ) {
    final var musics = playListService.getPlayListMusics(id, request.getMusicUserToken(), APPLE);
    final List<PlayListMusicResponse> response = PlayListMusicResponse.Companion.createList(musics);
    return BaseResponse.ok(response);
  }
}

File: src/main/kotlin/play/pluv/playlist/controller/PlayListController.kt
package play.pluv.playlist.controller

import jakarta.validation.Valid
import org.springframework.web.bind.annotation.*
import play.pluv.base.BaseResponse
import play.pluv.base.BaseResponse.Companion.ok
import play.pluv.playlist.application.PlayListService
import play.pluv.playlist.application.dto.*
import play.pluv.playlist.domain.MusicStreaming

@RestController
@RequestMapping("/playlist")
class PlayListController(
    private val playListService: PlayListService
) {

    @PostMapping("/spotify/read")
    fun readSpotifyPlayLists(
        @RequestBody request: @Valid PlayListReadRequest
    ): BaseResponse<List<PlayListOverViewResponse>> {
        val playLists = playListService.getPlayLists(
            accessToken = request.accessToken,
            source = MusicStreaming.SPOTIFY
        )
        val response = PlayListOverViewResponse.createList(playLists)
        return ok(response)
    }

    @PostMapping("/youtube/read")
    fun readYoutubePlayLists(
        @RequestBody request: @Valid PlayListReadRequest
    ): BaseResponse<List<PlayListOverViewResponse>> {
        val playLists = playListService.getPlayLists(
            accessToken = request.accessToken,
            source = MusicStreaming.YOUTUBE
        )
        val response = PlayListOverViewResponse.createList(playLists)
        return ok(response)
    }

    @PostMapping("/ocr/read")
    fun readOcrPlayLists(
        @RequestBody request: @Valid PlayListOcrRequest
    ): BaseResponse<List<PlayListMusicResponse>> {
        val musics = playListService.getOcrPlayListMusics(
            base64EncodedImages = request.base64EncodedImages
        )
        val responses = PlayListMusicResponse.createList(musics)
        return ok(responses)
    }

    @PostMapping("/spotify/{id}/read")
    fun readSpotifyMusics(
        @RequestBody request: @Valid PlayListReadRequest, @PathVariable id: String
    ): BaseResponse<List<PlayListMusicResponse>> {
        val musics = playListService.getPlayListMusics(
            playListId = id,
            accessToken = request.accessToken,
            source = MusicStreaming.SPOTIFY
        )
        val response = PlayListMusicResponse.createList(musics)
        return ok(response)
    }

    @PostMapping("/youtube/{id}/read")
    fun readYoutubeMusics(
        @RequestBody request: @Valid PlayListReadRequest, @PathVariable id: String
    ): BaseResponse<List<PlayListMusicResponse?>?> {
        val musics = playListService.getPlayListMusics(
            playListId = id,
            accessToken = request.accessToken,
            source = MusicStreaming.YOUTUBE
        )
        val response = PlayListMusicResponse.createList(musics)
        return ok(response)
    }

    @PostMapping("/apple/read")
    fun readApplePlayLists(
        @RequestBody request: @Valid ApplePlayListReadRequest
    ): BaseResponse<List<PlayListOverViewResponse>> {
        val playLists = playListService.getPlayLists(
            accessToken = request.musicUserToken,
            source = MusicStreaming.APPLE
        )
        val response = PlayListOverViewResponse.createList(playLists)
        return ok(response)
    }

    @PostMapping("/apple/{id}/read")
    fun readAppleMusics(
        @RequestBody request: @Valid ApplePlayListReadRequest, @PathVariable id: String
    ): BaseResponse<List<PlayListMusicResponse?>?> {
        val musics = playListService.getPlayListMusics(
            playListId = id,
            accessToken = request.musicUserToken,
            source = MusicStreaming.APPLE
        )
        val response = PlayListMusicResponse.createList(musics)
        return ok(response)
    }
}
